givenergy_modbus/client/client.py

```diff
@@ -78,23 +78,32 @@ async def connect(self) -> None:
     async def detect_plant(self, timeout: int = 1, retries: int = 3) -> None:
         """Detect inverter capabilities that influence how subsequent requests are made."""
         _logger.info("Detecting plant")
-        #print("Detecting plant")
-
+        from givenergy_modbus.model.inverter import Model
         # Refresh the core set of registers that work across all inverters
         #await self.refresh_plant(True, timeout=timeout, retries=retries)
+        
+        #Force 0x11 slave address only during detect
+        self.plant.slave_address=0x11
+        
         await self.refresh_plant(True, number_batteries=5, retries=3, timeout=1.0)
         #print("Plant Detected")
 
         # Use that to detect the number of batteries
         self.plant.detect_batteries()
         _logger.info("Batteries detected: %d", self.plant.number_batteries)
 
+        #set isAIO as needed
+        if self.plant.inverter.model == Model.ALL_IN_ONE:   # Check what other devices need 0x11
+            self.plant.slave_address = 0x11
+        else:
+            self.plant.slave_address = 0x31
+
         # Some devices support additional registers
         # When unsupported, devices appear to simple ignore requests
         possible_additional_holding_registers = [180, 240, 300, 360]
         for hr in possible_additional_holding_registers:
             try:
-                reqs = commands.refresh_additional_holding_registers(hr)
+                reqs = commands.refresh_additional_holding_registers(hr,self.plant.slave_address)
                 await self.execute(reqs, timeout=timeout, retries=retries)
```

Comment:
Not your change, but couldn't/shouldn't these probes be happening in parallel (by sending them all to a single exec() call), rather than sequential ?